[quant] Quantizable LSTM

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49866 [quant] Quantizable MultiheadAttention
* **#49671 [quant] Quantizable LSTM**

- Introduces the `torch.nn.quantizable` namespace
- Adds the `torch.nn.quantizable.LSTM` module

The point of the `quantizable` namespace is to segregate the purely quantized modules with the modules that could be quantized through a normal quantization flow, but are not using the quantized kernels explicitly.
That means the quantizable modules are functionally and numerically equivalent to the FP ones and can be used instead of the FP ones without any loss.

The main difference between the `torch.nn.LSTM` and the `torch.nn.quantizable.LSTM` is that the former one does not support observation for the linear layers, because all the computation is internal to the `aten` namespace.
The `torch.nn.quantizable.LSTM`, however, uses explicit linear layers that can be observed for further quantization.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_lstm
```

Differential Revision: [D25663870](https://our.internmc.facebook.com/intern/diff/D25663870)

Commit:
[quant] Quantizable LSTM

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable LSTM"




[ghstack-poisoned]

z-a-f commented:
The original was rewritten, but the comments are still valid: https://github.com/pytorch/pytorch/pull/46229

Commit:
Update on "[quant] Quantizable LSTM"




[ghstack-poisoned]

z-a-f commented:
@zrphercule Can you try out this PR and see if you get an error?

zrphercule commented:
No hay problema!

Zafar <notifications@github.com>于2020年12月21日 周一上午2:54写道：

> @zrphercule <https://github.com/zrphercule> Can you try out this PR and
> see if you get an error?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/49671#issuecomment-748910030>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFM6ZD2ELNYJ6NGC242NB3DSV4SN5ANCNFSM4VD5FFKA>
> .
>

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2734,6 +2734,109 @@ def test_qrnncell(self, num_batches, input_size, hidden_size, per_channel_quant)
 
 
+"""Tests recurrent layers."""
+class TestPTQRNN(TestCase):
```
can you merge this test case with TestQuantizedOps: https://github.com/pytorch/pytorch/blob/master/test/quantization/test_quantized_op.py#L124

jerryzh168 commented on test/test_quantization.py:
```diff
@@ -14,6 +14,7 @@
 from quantization.test_quantized_op import TestComparatorOps  # noqa: F401
 from quantization.test_quantized_op import TestPadding  # noqa: F401
 from quantization.test_quantized_op import TestQuantizedEmbeddingOps  # noqa: F401
```
I think we should merge these with TestQuantizedOps as well

jerryzh168 commented on torch/__init__.py:
```diff
@@ -574,6 +574,7 @@ def _assert(condition, message):
 import torch.futures
 import torch.nn
 import torch.nn.intrinsic
+import torch.nn.quantizable
```
should we put this under torch.nn.mo.quantizable since we are planning to move everything under mo?

z-a-f replied:
I was thinking about it -- but currently `mo` is in the `fb` namespace. Should I start introducing the new namespace?

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2734,6 +2734,109 @@ def test_qrnncell(self, num_batches, input_size, hidden_size, per_channel_quant)
+        # Moving the mean of the random input to avoid signal being close to 0
+        x = np.random.randn(seq_len, batch_size, input_size)
+        scale, zero_point = _calculate_dynamic_qparams(x, dtype=dtype)
+        qx = _quantize(x, scale=scale, zero_point=zero_point, dtype=dtype)
```
this is overwritten by L2787, do we need this?

z-a-f replied:
Oh, my bad!

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2734,6 +2734,109 @@ def test_qrnncell(self, num_batches, input_size, hidden_size, per_channel_quant)
+        dtype = np.uint8
+        qtype = torch.quint8
+
+        custom_module_config = {
```
should we add this support in the eager mode and fx graph mode flow by default instead of requiring user to do this every time?

raghuramank100 replied:
This is not done by default. For default users, we do module swapping and called the fused quantized lstm op. This is done only for cases when users need a lowered IR that contains the underlying linear ops.

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2734,6 +2734,109 @@ def test_qrnncell(self, num_batches, input_size, hidden_size, per_channel_quant)
+                self.assertEqual(y_ref, y)
+
+                # Quantize
+                lstm_quantized = torch.quantization.convert(lstm_prepared)
```
also, how is convert supported? I didn't see quantized lstm module, also there is no convert_custom_config_dict here.

z-a-f replied:
Currently, the convert treats the lstm as if it was just another module -- as long as it is quantizable, all of its layers are just replaced with the quantized ones. We cannot do it for prepare, because we need to avoid the insertion of the observers.

z-a-f commented on torch/nn/quantizable/modules/rnn.py:
```diff
@@ -0,0 +1,292 @@
+
+class LSTMCell(torch.nn.Module):
+    _FLOAT_MODULE = torch.nn.LSTMCell
+    # There are multiple outputs in the forward -- we don't want to observe!
+    do_not_observe: bool = True
```
It's introduced in the previous PR

raghuramank100 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2734,6 +2734,109 @@ def test_qrnncell(self, num_batches, input_size, hidden_size, per_channel_quant)
+            }
+        }
+
+        # Moving the mean of the random input to avoid signal being close to 0
```
Is this comment applicable now? The input is zero mean.

Commit:
Update on "[quant] Quantizable LSTM"


Differential Revision: [D25663870](https://our.internmc.facebook.com/intern/diff/D25663870)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable LSTM"


- Introduces the `torch.nn.quantizable` namespace
- Adds the `torch.nn.quantizable.LSTM` module

The point of the `quantizable` namespace is to segregate the purely quantized modules with the modules that could be quantized through a normal quantization flow, but are not using the quantized kernels explicitly.
That means the quantizable modules are functionally and numerically equivalent to the FP ones and can be used instead of the FP ones without any loss.

The main difference between the `torch.nn.LSTM` and the `torch.nn.quantizable.LSTM` is that the former one does not support observation for the linear layers, because all the computation is internal to the `aten` namespace.
The `torch.nn.quantizable.LSTM`, however, uses explicit linear layers that can be observed for further quantization.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_lstm
```

Differential Revision: [D25663870](https://our.internmc.facebook.com/intern/diff/D25663870)

[ghstack-poisoned]

vkuzo commented on torch/nn/quantizable/modules/rnn.py:
```diff
@@ -0,0 +1,288 @@
+into its building blocks to be able to observe.
+"""
+
+class LSTMCell(torch.nn.Module):
```
for all of these, can we add a `_get_name` method which includes the information that this is not the vanilla `nn.LSTMCell`, etc

vkuzo commented on torch/nn/quantizable/modules/rnn.py:
```diff
@@ -0,0 +1,288 @@
+            cell.hgates.bias = torch.nn.Parameter(bh)
+        return cell
+
+    @classmethod
+    def from_float(cls, other):
```
IMO it would be cleaner if this class had 0 references to quantization, and things like attaching the qconfig, etc, would be done by the user outside of this class.  This would remove the dependency of this module on quantization and make the code more modular.  What are your thoughts?

z-a-f replied:
That violates the whole quantization flow as we have it. The way we use both quantized and quantizable modules (custom or otherwise) is that we call the `from_float` on the FP module, which is expected to have the qconfig -- if it does not, that means the FP module should not be quantized. Also, the quantizable modules are not full replacement of the floating point ones -- they are designed to be used the same way the quantized ones are. The reason they are in a different namespace (as per discussion) is that this namespace does not have any quantized kernels, and can be run on the FP tensors directly.

vkuzo replied:
sg.  I think the `quantizable` naming is weird to think about, but makes sense, thanks.

vkuzo commented on torch/testing/_internal/common_quantized.py:
```diff
@@ -102,6 +102,27 @@ def _calculate_dynamic_per_channel_qparams(X, dtype):
 
     return scale, zero_point
 
+def _snr(x, x_hat):
```
nit: can we annotate types on function I/O

z-a-f replied:
This is a very good point. The short answer is **maybe**, but I don't see the reason why :)

The long answer is that the files under the `testing/_internal` are not expected to get jitted or to have a mandatory type checking. This means that other than readability the type hints in this folder will serve no purpose.

However, because mot functions under testing can handle multiple argument types, the readability will suffer a lot if type hints are added. For example, here how this function will look like:

```python
from typing import List, Tuple
from torch import Tensor

ListOfTensors = List[Union[List, Tuple, Tensor]]
TupleOfTensors = Tuple[Union[List, Tuple, Tensor]]

def _snr(x: Union[Tensor, ListOfTensors, TupleOfTensors],
         x_hat: Union[Tensor, ListOfTensors, TupleOfTensors]
) -> Union[Tensor, ListOfTensors, TupleOfTensors]:
    ...

# Or, if we are OK with being less descriptive:

def _snr(x: Union[Tensor, List, Tuple],
         x_hat: Union[Tensor, List, Tuple]
) -> Union[Tensor, List, Tuple]:
    ...

```

If you notice, the readability of the code does not improve, and because the function is not expected to get type checked, there are only downsides of annotating it. With that said, if we decide to annotate the rest of the code under the `testing/_internal`, I think we can start now

vkuzo replied:
> However, because mot functions under testing can handle multiple argument types, the readability will suffer a lot if type hints are added. For example, here how this function will look like:

The answer might vary per person, but having types significantly improves readability for me personally, even if it might decrease it for you.  Are there any compromises we can go for here?

z-a-f replied:
Wait, I thought I already added args description in the code to avoid the type annotations. Can you, please, refresh?

vkuzo replied:
```
ListOfTensors = List[Union[List, Tuple, Tensor]]
TupleOfTensors = Tuple[Union[List, Tuple, Tensor]]

def _snr(x: Union[Tensor, ListOfTensors, TupleOfTensors],
         x_hat: Union[Tensor, ListOfTensors, TupleOfTensors]
) -> Union[Tensor, ListOfTensors, TupleOfTensors]:
```

if you're up for doing this, that would be awesome.  I think it's much more readable than not having it.

vkuzo reviewed (commented):
makes sense, had some questions inline.  High level, can you comment on what are the performance implications of this in fp32, and why not just change the existing LSTM module.  I remember we talked about this some time back, but would be great to document the tradeoff here.

Commit:
Update on "[quant] Quantizable LSTM"


- Introduces the `torch.nn.quantizable` namespace
- Adds the `torch.nn.quantizable.LSTM` module

The point of the `quantizable` namespace is to segregate the purely quantized modules with the modules that could be quantized through a normal quantization flow, but are not using the quantized kernels explicitly.
That means the quantizable modules are functionally and numerically equivalent to the FP ones and can be used instead of the FP ones without any loss.

The main difference between the `torch.nn.LSTM` and the `torch.nn.quantizable.LSTM` is that the former one does not support observation for the linear layers, because all the computation is internal to the `aten` namespace.
The `torch.nn.quantizable.LSTM`, however, uses explicit linear layers that can be observed for further quantization.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_lstm
```

Differential Revision: [D25663870](https://our.internmc.facebook.com/intern/diff/D25663870)

[ghstack-poisoned]

vkuzo commented on torch/nn/quantizable/modules/rnn.py:
```diff
@@ -0,0 +1,294 @@
+        observed = torch.quantization.prepare(observed, inplace=True)
+        return observed
+
+    def from_observed(self, other):
```
I see the calls to this in the custom module code, but this PR seems like the first non-test use case.  Is that expected?  Should this be an interface with a docblock somewhere?

If this is the only callsite, should we even have this API at all?  Can we just inline the `torch.quantization.convert` call in the custom module flow instead, and have one less API surface to worry about?  cc @jerryzh168

z-a-f replied:
We had a minor discussion with @jerry39213gh about exactly that -- the reason we want to have the `from_observed` is because the conversion process is not necessarily producing structurally equivalent module. For example, if we decide to rewrite the LSTM in C++, and have a module that just calls it in the `forward`, we could. In such a case, the `from_float` would convert the FP module into the "quantizable" version that will be calibrated/observed, and later converted using the `from_observed`. In such (hypothetical) scenario, the `from_observed` would not call the `convert`, but explicitly remove the internal linear layers and replace all of them with a single call to an `aten` kernel.

As for the docs: This will be documented the same way the `from_float` is documented: Once the custom module is more or less stable, we will add a tutorial + a paragraph in the quantization documentation

vkuzo commented on torch/nn/quantizable/modules/rnn.py:
```diff
@@ -0,0 +1,294 @@
+class LSTMCell(torch.nn.Module):
+    _FLOAT_MODULE = torch.nn.LSTMCell
+
+    def __init__(self, input_dim, hidden_dim, bias=True):
```
can we make sure mypy is enabled for this module (in caffe2/mypy.ini), and annotate types on at least function I/O?   It would help with readability.

z-a-f replied:
SG!

Commit:
Update on "[quant] Quantizable LSTM"


- Introduces the `torch.nn.quantizable` namespace
- Adds the `torch.nn.quantizable.LSTM` module

The point of the `quantizable` namespace is to segregate the purely quantized modules with the modules that could be quantized through a normal quantization flow, but are not using the quantized kernels explicitly.
That means the quantizable modules are functionally and numerically equivalent to the FP ones and can be used instead of the FP ones without any loss.

The main difference between the `torch.nn.LSTM` and the `torch.nn.quantizable.LSTM` is that the former one does not support observation for the linear layers, because all the computation is internal to the `aten` namespace.
The `torch.nn.quantizable.LSTM`, however, uses explicit linear layers that can be observed for further quantization.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_lstm
```

Differential Revision: [D25663870](https://our.internmc.facebook.com/intern/diff/D25663870)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable LSTM"


- Introduces the `torch.nn.quantizable` namespace
- Adds the `torch.nn.quantizable.LSTM` module

The point of the `quantizable` namespace is to segregate the purely quantized modules with the modules that could be quantized through a normal quantization flow, but are not using the quantized kernels explicitly.
That means the quantizable modules are functionally and numerically equivalent to the FP ones and can be used instead of the FP ones without any loss.

The main difference between the `torch.nn.LSTM` and the `torch.nn.quantizable.LSTM` is that the former one does not support observation for the linear layers, because all the computation is internal to the `aten` namespace.
The `torch.nn.quantizable.LSTM`, however, uses explicit linear layers that can be observed for further quantization.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_lstm
```

Differential Revision: [D25663870](https://our.internmc.facebook.com/intern/diff/D25663870)

[ghstack-poisoned]
